### PR TITLE
Allow different ElementMapper types in ElementIndexScatterHandle.

### LINKED ISSUE #2095: compilation error with dune 2.7/master
Compilation is done with dune master/2.7 installed (needed fix in dune master). Also this compilation have most dune modules available.
Part of error meassage is:
```
/home/hnil/Documents/GITHUB/OPM/opm_source/devel_swift/opm-simulators/ebos/collecttoiorank.hh:336:63: error: no matching function for call to ‘Opm::CollectDataToIORank<Opm::EclCpGridVanguard<Opm::Properties::TTag::EbosTypeTag> >::ElementIndexScatterHandle<Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid> >, Dune::Impl::MCMGFailLayout> >::ElementIndexScatterHandle(EquilElementMapper&, ElementMapper&, std::vector<int>&)’
                 ElementIndexScatterHandle<EquilElementMapper> handle(equilElemMapper, elemMapper, localIdxToGlobalIdx_);
                                                               ^~~~~~
/home/hnil/Documents/GITHUB/OPM/opm_source/devel_swift/opm-simulators/ebos/collecttoiorank.hh:214:9: note: candidate: ‘Opm::CollectDataToIORank<Vanguard>::ElementIndexScatterHandle<Mapper>::ElementIndexScatterHandle(const Mapper&, const Mapper&, std::vector<int>&) [with Mapper = Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid> >, Dune::Impl::MCMGFailLayout>; Vanguard = Opm::EclCpGridVanguard<Opm::Properties::TTag::EbosTypeTag>]’
         ElementIndexScatterHandle(const Mapper& sendMapper, const Mapper& recvMapper, std::vector<int>& elementIndices)
         ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/hnil/Documents/GITHUB/OPM/opm_source/devel_swift/opm-simulators/ebos/collecttoiorank.hh:214:9: note:   no known conversion for argument 2 from ‘ElementMapper’ {aka ‘Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::Fem::GridPart2GridViewTraits<Dune::Fem::AdaptiveLeafGridPart<Dune::CpGrid, (Dune::PartitionIteratorType)4, false> > >, Dune::Impl::MCMGFailLayout>’} to ‘const Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid> >, Dune::Impl::MCMGFailLayout>&’
/home/hnil/Documents/GITHUB/OPM/opm_source/devel_swift/opm-simulators/ebos/collecttoiorank.hh:211:11: note: candidate: ‘Opm::CollectDataToIORank<Opm::EclCpGridVanguard<Opm::Properties::TTag::EbosTypeTag> >::ElementIndexScatterHandle<Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid> >, Dune::Impl::MCMGFailLayout> >::ElementIndexScatterHandle(const Opm::CollectDataToIORank<Opm::EclCpGridVanguard<Opm::Properties::TTag::EbosTypeTag> >::ElementIndexScatterHandle<Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid> >, Dune::Impl::MCMGFailLayout> >&)’
     class ElementIndexScatterHandle
           ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/hnil/Documents/GITHUB/OPM/opm_source/devel_swift/opm-simulators/ebos/collecttoiorank.hh:211:11: note:   candidate expects 1 argument, 3 provided
/home/hnil/Documents/GITHUB/OPM/opm_source/devel_swift/opm-simulators/ebos/collecttoiorank.hh:211:11: note: candidate: ‘Opm::CollectDataToIORank<Opm::EclCpGridVanguard<Opm::Properties::TTag::EbosTypeTag> >::ElementIndexScatterHandle<Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid> >, Dune::Impl::MCMGFailLayout> >::ElementIndexScatterHandle(Opm::CollectDataToIORank<Opm::EclCpGridVanguard<Opm::Properties::TTag::EbosTypeTag> >::ElementIndexScatterHandle<Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid> >, Dune::Impl::MCMGFailLayout> >&&)’
/home/hnil/Documents/GITHUB/OPM/opm_source/devel_swift/opm-simulators/ebos/collecttoiorank.hh:211:11: note:   candidate expects 1 argument, 3 provided
```

### PR DESCRIPTION
**Pull request by blattms**
This is needed to support dune-fem where the local mapper might
be
```C++
MultipleCodimMultipleGeomTypeMapper<GridView<GridPart2GridViewTraits<AdaptiveLeafGridPart<CpGrid, (PartitionIteratorType)4,
false> > >, Dune::Impl::MCMGFailLayout>
```
as opposed to the global one being
```C++
MultipleCodimMultipleGeomTypeMapper<GridView<DefaultLeafGridViewTraits<CpGrid>>, Dune::Impl::MCMGFailLayout>
```

Closes #2095.

### CODE REVIEW

**Comment by hnil** [comment]
This seem to fix the issue. The changes I needed to do in the cmake system is pushed to 
https://github.com/hnil/opm-simulators/tree/markus_fix, but this may have been fixed in other pullrequest? It just add defines in opm-simulators for all the dune-modules which is found, but is not tested for many combinations of modules.

File: ebos/collecttoiorank.hh [hunk 1]
```diff
@@ -207,11 +207,11 @@ public:
     };
 
     /// \brief Communication handle to scatter the global index
-    template<class Mapper>
+    template<class EquilMapper, class Mapper>
```
Comment (atgeirr) [inline ebos/collecttoiorank.hh]: I am a little confused about the name EquilMapper, would not SendMapper and RecvMapper be better?

Comment (blattms) [reply ebos/collecttoiorank.hh]: Naming is consistent with previous naming of the variables used further [down](/OPM/opm-simulators/blob/608b317ecccf49ac9c8d27847332f2709ace8cd0/ebos/collecttoiorank.hh#L322-L345) (grid, equilGrid. Equil indicates that it is for the global grid.

But I agree using send and receive in the constructor is not. What should we do?

**Comment by blattms** [comment]
@hnil Thanks part of the cmake changes were int the closed PR #2093, but not all (grid, alugrid, localfunctions). Maybe @GitPaean can take a look at your commit 21cb882f7c?

File: ebos/collecttoiorank.hh (hunk 1 above)
Comment (atgeirr) [reply ebos/collecttoiorank.hh]: Thanks for your comment. I think "equil" meaning the full, global grid is not very intuitive (unless you have been reading too much of the OPM code...), perhaps we should use the term "global" instead? (Although that means "all logical cartesian cells" in some contexts, I guess we should replace them with "logical cartesian" as well.)

**Comment by GitPaean** [comment]
I checked a little bit. It looks like only the version of dune-fem is needed to compile flow. the content of https://github.com/OPM/opm-simulators/pull/2094 along with this PR is enough to compile flow. 

Please let me know if things work for you differently, @hnil.

File: ebos/collecttoiorank.hh (hunk 1 above)
Comment (blattms) [reply ebos/collecttoiorank.hh]: Since renaming allocurrences would change other parts of the file, should I do that here or in another PR?

**Comment by blattms** [comment]
If this is not merged today it will not make it to release candidate 1.

**Comment by hnil** [comment]
Please merge. We have tested. Better dependency checks may come as separate pullrequests

File: ebos/collecttoiorank.hh (hunk 1 above)
Comment (atgeirr) [reply ebos/collecttoiorank.hh]: Another PR is fine. I'll run jenkins and merge as is if green.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by blattms** [comment]
No I get why I did not experience this when testing the release. The PR that introduced `ElementIndexScatterHandle` is reverted there since it is buggy. Therefore this will not be in the release and does not need to be either

**Pull request merged**